orchestration/service/versions.go

```diff
@@ -0,0 +1,159 @@
+// Copyright 2022 Namespace Labs Inc; All rights reserved.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+
+package service
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"sync"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"namespacelabs.dev/foundation/framework/kubernetes/kubedef"
+	"namespacelabs.dev/foundation/internal/artifacts/oci"
+	"namespacelabs.dev/foundation/internal/fnapi"
+	"namespacelabs.dev/foundation/orchestration/proto"
+	"namespacelabs.dev/foundation/schema"
+)
+
+const (
+	serverId                      = "0fomj22adbua2u0ug3og"
+	serverPkg  schema.PackageName = "namespacelabs.dev/foundation/orchestration/server"
+	toolPkg    schema.PackageName = "namespacelabs.dev/foundation/orchestration/server/tool"
+	serverName                    = "orchestration-api-server"
+
+	backgroundUpdateInterval = 30 * time.Minute
+	fetchLatestTimeout       = 30 * time.Second // can be generous, since we don't block in this.
+	cacheTimeout             = time.Minute
+)
+
+type versionChecker struct {
+	serverCtx context.Context
+
+	current *schema.Workspace_BinaryDigest
+
+	mu        sync.Mutex
+	pinned    []*schema.Workspace_BinaryDigest
+	fetchedAt time.Time
+}
+
+func newVersionChecker(ctx context.Context) *versionChecker {
+	current, err := getCurrentVersion(ctx)
+	if err != nil {
+		log.Fatalf("unable to compute current version: %v", err)
+	}
+
+	vc := &versionChecker{
+		serverCtx: ctx,
+		current:   current,
+	}
+
+	go func() {
+		for {
+			vc.updateLatest()
+
+			time.Sleep(backgroundUpdateInterval)
+		}
+	}()
+
+	return vc
+}
+
+func (vc *versionChecker) GetOrchestratorVersion() *proto.GetOrchestratorVersionResponse {
+	vc.mu.Lock()
+	defer vc.mu.Unlock()
+
+	// shedule a non-blocking update so that future calls get the latest version.
+	go vc.updateLatest()
```

Comment:
I wouldn't do this -- I would add a flag "skip_cache" that instructs the orchestrator to ignore its cache and fetch the latest version.